staging/src/k8s.io/kms/encryption/service.go

```diff
@@ -53,121 +63,165 @@ const (
+	return localKEKService
 }
 
-func (m *LocalKEKService) getTransformerForEncryption(uid string) (value.Transformer, *service.EncryptResponse, error) {
-	// Check if we have a local KEK
-	//	- If exists, use the local KEK for encryption and return
-	//  - Not exists, generate local KEK, encrypt with remote KEK,
-	//	store it in cache encrypt the data and return. This can be
-	// 	expensive but only 1 in N calls will incur this additional latency,
-	// 	N being number of times local KEK is reused)
-	m.remoteOnce.Do(func() {
-		m.localTransformerErr = wait.PollImmediateWithContext(context.Background(), localKEKGenerationPollInterval, localKEKGenerationPollTimeout,
-			func(ctx context.Context) (done bool, err error) {
-				key, err := generateKey(keyLength)
-				if err != nil {
-					return false, fmt.Errorf("failed to generate local KEK: %w", err)
-				}
-				block, err := aes.NewCipher(key)
-				if err != nil {
-					return false, fmt.Errorf("failed to create cipher block: %w", err)
-				}
-				transformer := aestransformer.NewGCMTransformer(block)
-
-				resp, err := m.remoteKMS.Encrypt(ctx, uid, key)
-				if err != nil {
-					klog.ErrorS(err, "failed to encrypt local KEK with remote KMS", "uid", uid)
-					return false, nil
-				}
-				if err = validateRemoteKMSResponse(resp); err != nil {
-					return false, fmt.Errorf("response annotations failed validation: %w", err)
-				}
-				m.remoteKMSResponse = copyResponseAndAddLocalKEKAnnotation(resp)
-				m.localTransformer = transformer
-				m.transformers.Add(base64.StdEncoding.EncodeToString(resp.Ciphertext), transformer)
-				return true, nil
-			})
-	})
-	return m.localTransformer, m.remoteKMSResponse, m.localTransformerErr
+// Run method creates a new local KEK  when the following thresholds are met:
+//   - the local KEK is used more often than keySuggestedUsage times or
+//   - the local KEK is older than a localExpiry.
+//
+// this method starts the controller and blocks until the context is cancelled.
+func (m *LocalKEKService) run(ctx context.Context) {
+	wait.UntilWithContext(ctx, func(ctx context.Context) {
+		lk := m.getLocalKEK()
+		// this is the first time the local KEK is generated
+		localKEKNotGenerated := lk.transformer == nil
+		// the local KEK is used more often than keySuggestedUsage times
+		localKEKUsageThresholdReached := lk.usage.Load() > m.keySuggestedUsage
+		// the local KEK is older than the expiry
+		localKEKExpired := m.clock.Now().After(lk.expiry)
+
+		if localKEKNotGenerated || localKEKUsageThresholdReached || localKEKExpired {
+			uid := string(uuid.NewUUID())
+			err := m.generateLocalKEK(ctx, uid, "")
+			if err == nil {
+				m.isReady.Store(true)
+				return
+			}
+			klog.V(2).ErrorS(err, "failed to generate local KEK", "uid", uid)
+			// if the local KEK is expired and we cannot generate a new one, we set
+			// isReady to false because we can no longer encrypt new data.
+			if localKEKExpired {
+				m.isReady.Store(false)
+			}
+		}
+	}, m.pollInterval)
```

Comment:
Wire the clock through to the wait logic (I believe this passes the same params as `UntilWithContext` but lets you tweak the clock/timer so that when you step the fake clock forward, it triggers the timer immediately):

```suggestion
	// same as wait.UntilWithContext but with a custom clock
	wait.BackoffUntil(func() {
		lk := m.getLocalKEK()
		// this is the first time the local KEK is generated
		localKEKNotGenerated := lk.transformer == nil
		// the local KEK is used more often than keySuggestedUsage times
		localKEKUsageThresholdReached := lk.usage.Load() > m.keySuggestedUsage
		// the local KEK is older than the expiry
		localKEKExpired := m.clock.Now().After(lk.expiry)

		if localKEKNotGenerated || localKEKUsageThresholdReached || localKEKExpired {
			uid := string(uuid.NewUUID())
			err := m.generateLocalKEK(ctx, uid, "")
			if err == nil {
				m.isReady.Store(true)
				return
			}
			klog.V(2).ErrorS(err, "failed to generate local KEK", "uid", uid)
			// if the local KEK is expired and we cannot generate a new one, we set
			// isReady to false because we can no longer encrypt new data.
			if localKEKExpired {
				m.isReady.Store(false)
			}
		}
	}, wait.NewJitteredBackoffManager(m.pollInterval, 0, m.clock), true, ctx.Done())
```